Compilation errors for Linux edrv in kernel >v4.15

**Bug Description:**
Stack version : v2.7.0
SHA : df70e53c688f56b1
Test case : Compile test
Managing Node : Linux edrv designs

**Test environment details:**
**_Linux edrv:_** i210
**_OS:_** Ubuntu 18.04
**_Kernel version:_** 4.16.7-rt1 

**Bug Summary:**
Linux edrv driver compilation results with errors

**Steps to reproduce:**
Compile the Linux edrv driver

**Error log:**
[ 97%] Generating oplki210mn/src/hrestimer-i210.c
[100%] Generating oplki210mn/oplki210mn.ko
openPOWERLINK-dir/drivers/linux/drv_kernelmod_edrv/build/oplki210mn/src/main.c: In function ‘powerlinkOpen’:
openPOWERLINK-dir/drivers/linux/drv_kernelmod_edrv/build/oplki210mn/src/main.c:313:5: error: implicit declaration of function ‘init_timer’; did you mean ‘init_timers’? [-Werror=implicit-function-declaration]
     init_timer(&heartbeatTimer_g);
     ^~~~~~~~~~
     init_timers
openPOWERLINK-dir/drivers/linux/drv_kernelmod_edrv/build/oplki210mn/src/main.c: In function ‘startHeartbeatTimer’:
openPOWERLINK-dir/drivers/linux/drv_kernelmod_edrv/build/oplki210mn/src/main.c:933:31: error: assignment from incompatible pointer type [-Werror=incompatible-pointer-types]
     heartbeatTimer_g.function = increaseHeartbeatCb;
                               ^
openPOWERLINK-dir/drivers/linux/drv_kernelmod_edrv/build/oplki210mn/src/main.c:934:21: error: ‘struct timer_list’ has no member named ‘data’
     heartbeatTimer_g.data = 0;
                     ^
cc1: some warnings being treated as errors
scripts/Makefile.build:324: recipe for target '/openPOWERLINK-dir/drivers/linux/drv_kernelmod_edrv/build/oplki210mn/src/main.o' failed
make[4]: *** [openPOWERLINK-dir/drivers/linux/drv_kernelmod_edrv/build/oplki210mn/src/main.o] Error 1
Makefile:1561: recipe for target '_module_/openPOWERLINK-drivers/linux/drv_kernelmod_edrv/build/oplki210mn' failed
make[3]: *** [_module_openPOWERLINK-dir/drivers/linux/drv_kernelmod_edrv/build/oplki210mn] Error 2
CMakeFiles/oplki210mn.dir/build.make:96: recipe for target 'oplki210mn/oplki210mn.ko' failed
make[2]: *** [oplki210mn/oplki210mn.ko] Error 2
CMakeFiles/Makefile2:99: recipe for target 'CMakeFiles/oplki210mn.dir/all' failed
make[1]: *** [CMakeFiles/oplki210mn.dir/all] Error 2
Makefile:129: recipe for target 'all' failed
make: *** [all] Error 2

--- a/drivers/linux/drv_kernelmod_edrv/main.c
+++ b/drivers/linux/drv_kernelmod_edrv/main.c
@@ -11,7 +11,7 @@ the openPOWERLINK kernel stack.
 *******************************************************************************/
 /*------------------------------------------------------------------------------
 Copyright (c) 2016, B&R Industrial Automation GmbH
-Copyright (c) 2017, Kalycito Infotech Private Limited
+Copyright (c) 2018, Kalycito Infotech Private Limited
 All rights reserved.
 
 Redistribution and use in source and binary forms, with or without
@@ -157,7 +157,11 @@ static int          sendAsyncFrame(unsigned long arg_p);
 static int          writeErrorObject(unsigned long arg_p);
 static int          readErrorObject(unsigned long arg_p);
 
+#if (LINUX_VERSION_CODE <= KERNEL_VERSION(4, 15, 0))
 static void         increaseHeartbeatCb(ULONG data_p);
+#else
+static void         increaseHeartbeatCb(struct timer_list* data_p);
+#endif
 static void         startHeartbeatTimer(ULONG timeInMs_p);
 static void         stopHeartbeatTimer(void);
 
@@ -310,7 +314,13 @@ static int powerlinkOpen(struct inode* pInode_p,
         return -ENOTTY;
     }
 
+#if (LINUX_VERSION_CODE <= KERNEL_VERSION(4, 15, 0))
     init_timer(&heartbeatTimer_g);
+#else
+    /* timer_setup API is used instead of init_timer as it is deprecated
+     * in the new kernel */
+    timer_setup(&heartbeatTimer_g, increaseHeartbeatCb, 0);
+#endif
 
     if (ctrlk_init(NULL) != kErrorOk)
     {
@@ -930,8 +940,11 @@ The function starts the timer used for updating the heartbeat counter.
 //------------------------------------------------------------------------------
 void startHeartbeatTimer(ULONG timeInMs_p)
 {
+#if (LINUX_VERSION_CODE <= KERNEL_VERSION(4, 15, 0))
     heartbeatTimer_g.function = increaseHeartbeatCb;
     heartbeatTimer_g.data = 0;
+#endif
+
     heartbeatTimer_g.expires = jiffies + (timeInMs_p * HZ / 1000);
     add_timer(&heartbeatTimer_g);
 }
@@ -962,7 +975,12 @@ heartbeat counter.
 \ingroup module_driver_linux_kernel
 */
 //------------------------------------------------------------------------------
+#if (LINUX_VERSION_CODE <= KERNEL_VERSION(4, 15, 0))
 void increaseHeartbeatCb(ULONG data_p)
+#else
+/* Modified increaseHeartbeatCb to handle callback functionality */
+void increaseHeartbeatCb(struct timer_list* data_p)
+#endif
 {
     UNUSED_PARAMETER(data_p);
 

--- a/stack/src/kernel/event/eventkcal-linuxkernel.c
+++ b/stack/src/kernel/event/eventkcal-linuxkernel.c
@@ -14,6 +14,7 @@ kernelspace platform. It uses the circular buffer interface for all event queues
 
 /*------------------------------------------------------------------------------
 Copyright (c) 2016, B&R Industrial Automation GmbH
+Copyright (c) 2018, Kalycito Infotech Private Limited
 All rights reserved.
 
 Redistribution and use in source and binary forms, with or without
@@ -55,6 +56,11 @@ SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 #include <linux/sched.h>
 #include <linux/cpumask.h>
 #include <linux/uaccess.h>
+#if (LINUX_VERSION_CODE >= KERNEL_VERSION(4, 15, 0))
+/* Included types.h header to link sched_param structure */
+#include <linux/sched/types.h>
+#endif
+
 #include <asm/uaccess.h>
 #include <asm/atomic.h>
 

